docs/testing-principles.md

```diff
@@ -0,0 +1,200 @@
+we use high-level APIs like `task()` instead of low level ones like
+`task._run_task()`
+
+That said, a system as complicated as Cumulus Framework cannot be tested only
+from the _end_ user's perspective. Cumulus Framework consists of many components.
+
+The "users" of these components are programmers on our team, and they should
+be well-tested, just as if they were documented external APIs.
+
+For example, the YAML merge algorithm is a component. The XML manipulator is
+a component. The package XML library SHOULD BE a component, etc. These should
+have test cases that ensure that their interfaces do not accidentally change.
+
+## The Hard Cases: The Awkward Squad
+
+In the lingo of functional programmers, our tests doesn't deal with the
+"awkward squad" of issues other tests need to deal with:
+
+-   a variety of org shapes
+-   error/exception conditions
+-   mutable state (in the org)
+-   delays, timestamps and other temporal aspects
+
+"Old fashioned" data faking may be needed to deal with these. The rest
+of this document outlines techninques for handling these hard cases.
+
+### Org shapes
+
+Cumulus Framework has testing infrastructure which can create custom
+orgs for tests:
+
+```
+@pytest.mark.vcr()
+@pytest.mark.org_shape("person_accounts", "config_person_accounts")
+def test_some_person_account_feature(self):
+    ...
+```
+
+This will run the `config_person_accounts` flow against the `person_accounts`
+org shape and run the test in the context of that org. At least that's
+the goal. This infrastructure has not been tested or used much yet (e.g. for
+actual person accounts testing).
+
+### Simulating Networking Errors with VCR
+
+It is possible to simulate networking errors by hand-editing VCR files.
+This pattern can be observed [here](https://github.com/SFDO-Tooling/CumulusCI/blob/35b07ebfa80170f6e0e9d8cabd3da75bf41d5d59/cumulusci/utils/salesforce/tests/test_count_sobjects.py#L23) and [here](https://github.com/SFDO-Tooling/CumulusCI/blob/e3ba7e8b542755b6395dbc895daee3d1ea3d11dc/cumulusci/utils/salesforce/tests/cassettes/ManualEdit_TestCountSObjects.test_count_sobjects__network_errors.yaml#L16).
+
+### Service Connectors
+
+Third party services (including Salesforce) should always be wrapped in Service Connector
+abstractions.
+
+These abstractions can be tested with Requests and VCR.
+
+The overall business logic can be tested by injecting fakes for Service Connectors.
+
+A client library like simple_salesforce or github3 is not necessarily a good service
+connector abstraction, because they work at a protocol level and not a business logic
+level. Two examples, from simple_salesforce:
+
+1. We make heavy use of simple_salesforce.restful(), which tunnels MANY different types
+   of requests through a single endpoint and is therefore hard to fake.
+
+2. We make heavy use of simple_salesforce.query() which has the same problem.
+
+Github3 is better, but arguably if we had started using it by wrapping it we'd be in a
```

Comment:
```suggestion
`Github3` is better, but arguably if we had started using it by wrapping it we'd be in a
```

Reply:
This whole paragraph seems like an explanation of implementation thought process. Do we need it? If so, it's a little hard to follow.